fix(core): deprecate non-Expr args in Add/Mul/Pow

oscarbenjamin opened the pull request:
Deprecates creating an Add, Mul or Pow with non-expr args e.g.:

    Add(Tuple(1, 2), FiniteSet(2), S.true, 0)

<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->

Related issues and PRs #19429, #5031, #18053, #18613

#### Brief description of what is fixed or changed


#### Other comments


#### Release Notes

<!-- Write the release notes for this release below. See
https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more information
on how to write release notes. The bot will check your release notes
automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* core
    * **DEPRECATION**: Using non-Expr args in Add, Mul or Pow is now deprecated
<!-- END RELEASE NOTES -->

oscarbenjamin commented on sympy/core/exprtools.py:
```diff
@@ -1316,22 +1316,6 @@ def _mask_nc(eq, name=None):
     >>> _mask_nc(eq, 'd')
     (A*_d0 + B*_d1, {_d0: Commutator(A, B), _d1: Commutator(A, C)}, [_d0, _d1, A, B])
 
-    If there is an object that:
-
-        - doesn't contain nc-symbols
-        - but has arguments which derive from Basic, not Expr
-        - and doesn't define an _eval_is_commutative routine
-
-    then it will give False (or None?) for the is_commutative test. Such
-    objects are also removed by this routine:
-
-    >>> from sympy import Basic
-    >>> eq = (1 + Mul(Basic(), Basic(), evaluate=False))
```
@smichr this doctest failed because in this PR it is not possible to create a `Mul` with non-`Expr` args.

Can you think of another example where this would apply or is it okay just to remove this as unneeded now?

Maybe the accompanying function could also be simplified after this...

sylee957 commented on sympy/core/operations.py:
```diff
@@ -25,16 +25,18 @@ class AssocOp(Basic):
     # and keep it right here
     __slots__ = ('is_commutative',)  # type: Tuple[str, ...]
 
+    _args_type = None
```
Why not make this `Basic`?

sylee957 replied:
Okay, it can also be redundant to check against basic here.

oscarbenjamin committed:
fix(core): deprecate non-Expr args in Add/Mul/Pow

Give a DeprecationWarning when creating an Add/Mul/Pow with non-Expr
args e.g.:

    Add(Tuple(1, 2), FiniteSet(2), S.true, 0)

oscarbenjamin commented:
This is ready for review

smichr commented:
Should args to defined functions be Expr, too?
```python
>>> f=Function('f')
>>> f(Basic())  #Undefined
f(Basic())
>>> f(Basic())+f(Basic())
2*f(Basic())
>>> sin(Basic())  # defined
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "\sympy\core\cache.py", line 94, in wrapper
    retval = cfunc(*args, **kwargs)
  File "\sympy\core\function.py", line 465, in __new__
    result = super().__new__(cls, *args, **options)
  File "\sympy\core\cache.py", line 94, in wrapper
    retval = cfunc(*args, **kwargs)
  File "\sympy\core\function.py", line 280, in __new__
    evaluated = cls.eval(*args)
  File "\sympy\functions\elementary\trigonometric.py", line 297, in eval
    if arg.could_extract_minus_sign():
AttributeError: 'Basic' object has no attribute 'could_extract_minus_sign'
```

oscarbenjamin commented:
In principle there can be defined functions that would have non-Expr args but I don't know of any in sympy.

Certainly functions like `sin` etc should have a way of checking that they receive sensible input. It should be more restrictive than just `Expr` though. For `sin` we really need the input to be something representing a numeric expression and not a vector, matrix, quantum operator etc. There are a few cases where defined functions accept things like matrices e.g.:
```julia
In [10]: exp(Matrix([[0, 2], [-2, 0]]))                                                                                           
Out[10]: 
⎡cos(2)   sin(2)⎤
⎢               ⎥
⎣-sin(2)  cos(2)⎦
```
That sort of thing should be discouraged though. There is the `Matrix.exp()` method for explicit matrices. A better way is to have a separate MatrixExp class that can also handle matrix symbols and things like this:
```julia
In [11]: M = MatrixSymbol('M', 2, 2)                                                                                              

In [12]: e = exp(M*t)                                                                                                             

In [13]: e                                                                                                                        
Out[13]: 
 t⋅M
ℯ   

In [14]: e.diff(t)
...
NotImplementedError: noncommutative scalars in MatMul are not supported.
```

smichr commented on sympy/printing/pretty/pretty.py:
```diff
@@ -1933,8 +1933,7 @@ def _print_Fraction(self, expr):
         if len(p.sets) >= 1 and not has_variety(p.sets):
-            from sympy import Pow
-            return self._print(Pow(p.sets[0], len(p.sets), evaluate=False))
+            return self._print(p.sets[0]) ** self._print(len(p.sets))
```
Why did this need to be changed?

oscarbenjamin replied:
We can't use `p.sets[0]` which is a `Set` in the args of a `Pow`.

smichr replied:
Oh..I get it. It is now a Set "Pow"  not and Expr Pow.

oscarbenjamin commented:
Thanks @smichr!
